File: frontend/src/stores/simulation.js
import { defineStore } from 'pinia'
import axios from 'axios'

const API = '/api'

export const useSimulationStore = defineStore('simulation', {
  state: () => ({
    cohortId: null,
    cohortMeta: null,
    simulationId: null,
    simulationStatus: null,   // running | completed | failed
    progress: 0,
    totalAgents: 0,
    completedAgents: 0,
    statistics: null,
    populationAnalysis: null,
    agentResults: [],
    trialConfig: null,
    error: null,
    loading: false,
    drugs: [],
    populations: [],
    pollTimer: null,
  }),

  actions: {
    async fetchDrugs() {
      const { data } = await axios.get(`${API}/drugs`)
      this.drugs = data.drugs
    },

    async fetchPopulations() {
      const { data } = await axios.get(`${API}/populations`)
      this.populations = data.populations
    },

    async uploadCohort(file, drug, dose_mg) {
      this.loading = true
      this.error = null
      try {
        const form = new FormData()
        form.append('file', file)
        form.append('drug', drug)
        form.append('dose_mg', dose_mg)
        const { data } = await axios.post(`${API}/cohort/upload`, form)
        this.cohortId = data.cohort_id
        this.cohortMeta = data
        return data
      } catch (e) {
        this.error = e.response?.data?.error || e.message
        throw e
      } finally {
        this.loading = false
      }
    },

    async createSyntheticCohort({ drug, dose_mg, n_agents, population, seed }) {
      this.loading = true
      this.error = null
      try {
        const { data } = await axios.post(`${API}/cohort/synthetic`, {
          drug, dose_mg, n_agents, population, seed
        })
        this.cohortId = data.cohort_id
        this.cohortMeta = data
        return data
      } catch (e) {
        this.error = e.response?.data?.error || e.message
        throw e
      } finally {
        this.loading = false
      }
    },

    async startSimulation(trialConfig) {
      this.loading = true
      this.error = null
      this.simulationStatus = null
      this.statistics = null
      this.populationAnalysis = null
      this.agentResults = []
      try {
        const { data } = await axios.post(`${API}/simulation/run`, {
          cohort_id: this.cohortId,
          ...trialConfig,
        })
        this.simulationId = data.simulation_id
        this.simulationStatus = 'running'
        this.totalAgents = data.total_agents
        this.trialConfig = data.trial_config
        this._startPolling()
        return data
      } catch (e) {
        this.error = e.response?.data?.error || e.message
        this.simulationStatus = 'failed'
        throw e
      } finally {
        this.loading = false
      }
    },

    _startPolling() {
      if (this.pollTimer) clearInterval(this.pollTimer)
      this.pollTimer = setInterval(async () => {
        await this._pollStatus()
      }, 2000)
    },

    async _pollStatus() {
      if (!this.simulationId) return
      try {
        const { data } = await axios.get(`${API}/simulation/status/${this.simulationId}`)
        this.simulationStatus = data.status
        this.progress = data.progress_pct
        this.completedAgents = data.completed_agents
        this.totalAgents = data.total_agents

        if (data.status === 'completed') {
          clearInterval(this.pollTimer)
          await this._fetchResults()
        } else if (data.status === 'failed') {
          clearInterval(this.pollTimer)
          this.error = data.error || 'Simulation failed'
        }
      } catch (e) {
        // ignore transient poll errors
      }
    },

    async _fetchResults() {
      const { data } = await axios.get(`${API}/simulation/result/${this.simulationId}`)
      this.statistics = data.statistics
      this.populationAnalysis = data.population_analysis
      this.trialConfig = data.trial_config

      const agentsResp = await axios.get(`${API}/simulation/result/${this.simulationId}/agents?limit=200`)
      this.agentResults = agentsResp.data.results
    },

    reset() {
      if (this.pollTimer) clearInterval(this.pollTimer)
      this.cohortId = null
      this.cohortMeta = null
      this.simulationId = null
      this.simulationStatus = null
      this.progress = 0
      this.statistics = null
      this.populationAnalysis = null
      this.agentResults = []
      this.error = null
    }
  }
})

File: backend/app/api/simulation.py
"""
Simulation API — trigger runs, poll status, fetch results.
"""

from __future__ import annotations

import uuid
from flask import Blueprint, request, jsonify, current_app

from app.services.swarm.simulation_engine import SimulationEngine, get_simulation, list_simulations
from app.utils.logger import get_logger

simulation_bp = Blueprint("simulation", __name__)
logger = get_logger("genomicswarm.api.simulation")

_engine: SimulationEngine | None = None


def get_engine() -> SimulationEngine:
    global _engine
    if _engine is None:
        api_key = current_app.config.get("ANTHROPIC_API_KEY", "")
        model = current_app.config.get("CLAUDE_MODEL", "claude-sonnet-4-6")
        workers = current_app.config.get("SIM_WORKERS", 10)
        _engine = SimulationEngine(api_key=api_key, model=model, max_workers=workers)
    return _engine


@simulation_bp.route("/run", methods=["POST"])
def run_simulation():
    """
    Start a simulation run.
    Body JSON:
    {
        "cohort_id": "<uuid>",           # from cohort upload
        "drug": "codeine",
        "dose_mg": 30,
        "route": "oral",
        "duration_days": 14,
        "indication": "pain management",
        "comparator": "placebo",
        "max_agents": 100               # optional override
    }
    Returns { simulation_id } immediately; poll /status/<id> for progress.
    """
    from app.api.cohort import get_cohort_personas

    data = request.get_json() or {}
    cohort_id = data.get("cohort_id")

    if not cohort_id:
        return jsonify({"error": "cohort_id required"}), 400

    personas = get_cohort_personas(cohort_id)
    if personas is None:
        return jsonify({"error": f"Cohort '{cohort_id}' not found"}), 404

    max_agents = min(
        int(data.get("max_agents", len(personas))),
        current_app.config.get("MAX_AGENTS", 1000),
    )
    personas = personas[:max_agents]

    trial_config = {
        "drug": data.get("drug", "codeine"),
        "dose_mg": float(data.get("dose_mg", 30.0)),
        "route": data.get("route", "oral"),
        "duration_days": int(data.get("duration_days", 14)),
        "indication": data.get("indication", "unspecified"),
        "comparator": data.get("comparator", "placebo"),
    }

    sim_id = str(uuid.uuid4())
    engine = get_engine()
    engine.run_async(sim_id, personas, trial_config)

    logger.info(f"Simulation {sim_id} started: {len(personas)} agents, {trial_config['drug']}")
    return jsonify({
        "simulation_id": sim_id,
        "status": "running",
        "total_agents": len(personas),
        "trial_config": trial_config,
    }), 202


@simulation_bp.route("/status/<sim_id>", methods=["GET"])
def simulation_status(sim_id: str):
    """Poll simulation progress."""
    sim = get_simulation(sim_id)
    if not sim:
        return jsonify({"error": "Simulation not found"}), 404

    completed = sim.get("completed_agents", 0)
    total = sim.get("total_agents", 1)
    progress_pct = round(100 * completed / total) if total else 0

    return jsonify({
        "simulation_id": sim_id,
        "status": sim.get("status"),
        "progress_pct": progress_pct,
        "completed_agents": completed,
        "total_agents": total,
        "started_at": sim.get("started_at"),
        "completed_at": sim.get("completed_at"),
        "error": sim.get("error"),
        # Include stats if done
        "statistics": sim.get("statistics") if sim.get("status") == "completed" else None,
    })


@simulation_bp.route("/result/<sim_id>", methods=["GET"])
def simulation_result(sim_id: str):
    """Fetch full simulation result (once completed)."""
    sim = get_simulation(sim_id)
    if not sim:
        return jsonify({"error": "Simulation not found"}), 404
    if sim.get("status") != "completed":
        return jsonify({"error": "Simulation not yet complete", "status": sim.get("status")}), 202

    # Return result without raw agent_results blob (use /agents for that)
    return jsonify({
        "simulation_id": sim_id,
        "status": sim["status"],
        "trial_config": sim["trial_config"],
        "statistics": sim["statistics"],
        "population_analysis": sim["population_analysis"],
        "started_at": sim["started_at"],
        "completed_at": sim["completed_at"],
        "total_agents": sim["total_agents"],
    })


@simulation_bp.route("/result/<sim_id>/agents", methods=["GET"])
def simulation_agents(sim_id: str):
    """Fetch paginated agent-level results."""
    sim = get_simulation(sim_id)
    if not sim:
        return jsonify({"error": "Simulation not found"}), 404

    results = sim.get("agent_results", [])
    limit = int(request.args.get("limit", 100))
    offset = int(request.args.get("offset", 0))

    return jsonify({
        "total": len(results),
        "results": results[offset:offset + limit],
    })


@simulation_bp.route("/list", methods=["GET"])
def list_sims():
    """List all simulations (summaries only)."""
    return jsonify({"simulations": list_simulations()})
